feat: implement responsiveness with srcSet and sizes

frederickfogerty opened the pull request:
## Description

This PR fundamentally changes the approach taken to responsiveness. Previously, the image (or component) was rendered with an empty image, and the size of the resulting DOM element was measured. An image was loaded with these dimensions, ensuring that the image loaded was only as large as needed. Now, this component achieves the same behaviour by leveraging new native browser features, srcSet and sizes. This allows the browser to load an appropriate size based on the expected image size, which is calculated based on sizes. This has an added benefit that images can load much quicker as the browser can load them as soon as page layout has completed, or sooner (e.g. if sizes is 100vw). This will work well with Fastboot.

The significant amount of breaking changes makes sense due to a lot of the props no longer making sense.

## TODO

- [x] README
- [ ] ~Update dummy app examples - after offline discuss about aspect ratio~
- [ ] ~maintaining aspect ratio with dpr~
- [x] add disableSrcSet attribute
- [x] remove didUpdateAttrs
- [x] remove "sizes is required"

### New Feature

- [x] If this is a big feature with breaking changes, consider [opening an issue][issues] to discuss first. This is completely up to you, but please keep in mind that your PR might not be accepted.
- [x] Run unit tests to ensure all existing tests are still passing
- [x] Add new passing unit tests to cover the code introduced by your PR
- [x] Update the readme
- [x] Add some [steps](#steps-to-test) so we can test your cool new feature!
- [x] The PR title is in the [conventional commit](#conventional-commit-spec) format: e.g. `feat(<area>): added new way to do this cool thing #issue-number`
- [x] Add your info to the [contributors](#packagejson-contributors) array in package.json!

## Steps to Test

Review the changes to the tests. The examples can also be served by cloning this PR, running `npm install` and then `ember serve`

BREAKING CHANGES:

- srcSet behaviour has changed to use sizes + srcSets
- the following attributes have been removed: pixelStep, auto

jayeb commented on README.md:
```diff
@@ -6,6 +6,14 @@ An Ember addon for easily adding responsive imagery via [imgix](https://www.imgi
 
+## Overview / Resources
+
+**Before you get started with ember-cli-imgix**, it's _highly recommended_ that you read Eric Portis' [seminal article on `srcset` and `sizes`](https://ericportis.com/posts/2014/srcset-sizes/). This article explains the history of responsive images in responsive design, why they're necessary, and how all these technologies work together to save bandwidth and provide a better experience for users. ...
```
A "react-imgix" snuck through here.

frederickfogerty replied:
Whoops!

jayeb commented on README.md:
````diff
@@ -42,18 +50,33 @@ module.exports = function(environment) {
+**Width and height known:** If the width and height are known beforehand, it is recommended that they are set explicitly:
+
+```hbs
+{{imgix-image path='/users/1.png' sizes='100vw' width=100 height=100}}
````
`sizes` shouldn't be required if both `width` and `height` are specified, right? If they're both provided, any reason we can't just automatically generate a DPR-style `srcset` and forget about `sizes`?

jayeb commented on README.md:
```diff
@@ -184,6 +205,17 @@ import ImgixCoreJs from 'imgix-core-js';
+
+## Browser Support
+
+- By default, browsers that don't support [`srcset`](http://caniuse.com/#feat=srcset), [`sizes`](http://caniuse.com/#feat=srcset), or [`picture`](http://caniuse.com/#feat=picture) will gracefully fall back to the default `img` `src` when appropriate. If you want to provide a fully-responsive experience for these browsers, react-imgix works great alongside [Picturefill](https://github.com/scottjeh ...
```
Another "react-imgix" snuck in here.

jayeb commented on addon/components/imgix-image.js:
```diff
@@ -73,34 +51,10 @@ export default Component.extend(ResizeAware, {
-    );
   },
 
   didUpdateAttrs(...args) {
```
If all this is gonna do is call `super`, can we remove the definition entirely?

frederickfogerty replied:
@Duder-onomy are we able to do this?

Duder-onomy replied:
Yeah, we can eliminate that hook.

Duder-onomy commented:
I pulled in this branch, many things seem to be diff. Just a quick glance at the examples page and most look broken. 

- Because we no longer use the actual rendered sizes, the aspect ration param no longer works. 

- We seem to have lost support for passing explicit width/height to the component. 

---

But more importantly, and you would know better than I, but how/why are most people using imgix? My assumption was that the vast majority use it for:
- To normalize user submitted image content
- To provide a fast resize and cache solution to avoid having to care about srcsets etc. 

I can confidently say that at TheDyrt (and all times I have used imgix in the past on diff projects), we use it for both of the above. Delivering to the client only the pixels we need. No more, no less. We use srcsets for image assets that we can control and predict all the possible sizes we might need (background/hero images for example). 

This WOULD help us in fastboot as we cannot know the viewport size and could be a viable solution for all, but maybe I am missing something important...

---

Looking at example page: (maybe it just needs to be updated?)
- The srcset created is incredibly verbose. 31 entries for a simple image. 
- The final rendered image does not display correctly without css to position it. For example: 
![image](https://user-images.githubusercontent.com/1612785/44681939-dc6d4e00-a9f6-11e8-89da-5e07cc08aca7.png)

---

Not trying to push back too hard, but wondering if we should support both srcset style and actual rendered pixel style?

frederickfogerty commented:
Hey there @Duder-onomy! Thank you for the comments. I think the best way to address your discussion is to respond to it inline.

> I pulled in this branch, many things seem to be diff. Just a quick glance at the examples page and most look broken.
> - Because we no longer use the actual rendered sizes, the aspect ratio param no longer works.

My apologies for not calling this out, this is actually still TODO. I realised that currently, this library supports the aspectRatio param, but with the new implementation this was initially not possible. imgix is looking to support aspect ratio in the future, but I think I may have a way to support it with the current API. I'm going to discuss with imgix engineers today. This should allow us to render these images correctly.

> - We seem to have lost support for passing explicit width/height to the component.

This is not intentional, I'll have a look at this today.

> But more importantly, and you would know better than I, but how/why are most people using imgix? My assumption was that the vast majority use it for:
> - To normalize user submitted image content
> - To provide a fast resize and cache solution to avoid having to care about srcsets etc.
> 
> I can confidently say that at TheDyrt (and all times I have used imgix in the past on diff projects), we use it for both of the above. Delivering to the client only the pixels we need. No more, no less. We use srcsets for image assets that we can control (background/hero images for example).
> 
> This WOULD help us in fastboot as we cannot know the viewport size, but maybe I am missing something important...

I don't quite know the best way to answer this, as I'm not sure of the type of behaviour that is expected.

The "only the pixels we need" is still possible in this approach. At imgix the rationale behind this is that we believe the following is true for most developers and images:

- the developer knows the exact right dimensions of the image to be displayed. In this case, the developer can specify those as `width`, `height`, and `sizes`, and that will be the size of the images passed to the client. 2x and 3x will be handled by the browser by using the srcset + sizes.
- the developer wants the image to resize dynamically based on screen width. This is possible by specifying `sizes` and letting the browser choose the right srcset based on the width. 

> Looking at example page: (maybe it just needs to be updated?)
>
> The srcset created is incredibly verbose. 31 entries for a simple image. 
> The final rendered image does not display correctly without css to position it. For example:

Agreed, the srcset is verbose, but it covers all possible physical resolutions from 100px to 8192px. We haven't found any downside to including this many srcsets.

What is meant by it doesn't display correctly? At the moment the squishing behaviour is not desirable, but we can hopefully fix this by adding the ability to maintain the aspect ratio.

> Not trying to push back too hard, but wondering if we should support both srcset style and actual rendered pixel style?

Ultimately we want to find something that works for most people. We've implemented this across most of our libraries, and it's been out in imgix.js for about a year now, so this is definitely not something new for us. 

A problem with the old approach is that browser support for it was buggy to say the least (at least over at react-imgix). This approach has the benefit of being a browser standard, and also works with server rendering. It takes away a specific use case where the developer has no idea what size the image is going to be rendered at. With imgix.js we removed support for this and faced no significant backlash. At react-imgix we supported this with an example, but didn't bake it in the core.

I'm keen to hear your thoughts on my responses above. We want to create something that works for everyone (or most people).

jayeb commented:
I would add that another problem with the old approach, and the primary reason we've moved away from this approach in other projects (most notably [imgix.js](https://github.com/imgix/imgix.js) and [react-imgix](https://github.com/imgix/react-imgix)) is that while the "try to serve the exact right number of pixels" approach is noble, you can end up doing your users a disservice by caching far too many image variants.

The goal shouldn't actually be to serve any given user a perfectly-sized image. The goal should be to serve them an image that:
1) is close enough to perfectly-sized that any artifacts from up- or down-scalable aren't noticeable, and
2) is already cached at the CDN level.

Through some testing we've determined that the optimal way to do this is to generate a robust (but not insane) set of URLs and let the browser decide which is the best one given the `sizes` attribute and the specifics of the user's device. This still leverages imgix's core strength of being able to easily generate derivatives of a master image at any conceivable width, but also takes into account the practical consideration of how to best deliver those images to the end-user: through a toasty-warm cache.

Duder-onomy commented:
Thanks for your thoughts on all this. 
I agree that this would be a improvement for the general use case. 

I am all for you merging it and this being the way the official imgix lib works (so it works like other imgix libs). I totally agree with using srcsets for images as often as possible. This will allow me to use imgix with other images on our site for which we have rolled our own resize/cache service to populate a srcset. 

^ pulling these changes into our app breaks most images (incorrect displayed size, odd overflows / stretching ). This breakage is mostly because we treat our imgix component like "Here is a hole in the DOM, fill it with an image perfectly sized and cropped" instead of, "it will be image like, and needs to have appropriate styling to prevent stretch and to  overflow only the edges we want". 

For thedyrt.com, I will not be able to use the srcset implementation above for most of our images. This does not break user expectation as the variance in requested sizes is actually not as big as one would expect.  So I feel like, we can have a sep component to accomplish thedyrt's use case (maybe not typical? so maybe something that ember-cli-imgix should not support?) OR I will need to maintain and support something else that accomplishes this. I am good either way.

frederickfogerty commented:
Hey @Duder-onomy thanks for all your thoughts. Glad to see that you see the improvements from these changes. 

I imagined that it would probably be a big breaking change for your application. Over at react-imgix, we wanted to help the people that were using it in the same way that you are, but we didn't want to support an official version. Thus, we made an example to show how a user could use the component in the old way. [The example can be found here.](https://github.com/imgix/react-imgix/blob/master/examples/fit-to-size-of-container.md) I'm happy to work together to publish a similar example here.

frederickfogerty commented:
Regarding aspect ratio, I think the best thing is to merge this PR into a v2 branch, and wait until aspect ratio is implemented by Imgix to cut a release so that we don't remove aspect ratio support, to only add it back in again.

@jayeb this is ready for a final review, although I haven't changed much since the last review

Duder-onomy commented:
Cool. I agree with this merge. srcsets and sizes are def the way to go with web dev these days. 

Regarding my current use case, ( and I still suspect the primary reason people are using imgix ),  I will probably end up maintaining a fork again, something like ember-imgix-fit-to-container. Let me know if it is something you want to be a part of. The [example from react imgix](https://github.com/imgix/react-imgix/blob/master/examples/fit-to-size-of-container.md) is super close to what I want less a few important bits, like aspect ratio and traversing parent elements to measure. When that work is done you can just point people that way if they want that behavior.

frederickfogerty commented on addon/components/imgix-image.js:
```diff
@@ -180,16 +136,48 @@ export default Component.extend(ResizeAware, {
+              ...options,
+              dpr
+            });
+          srcSet = `${buildWithDpr(2)} 2x, ${buildWithDpr(3)} 3x`;
```
Are there actually any devices that are 4x or 5x? I couldn't find any with a quick Google search. If there are, then we should add 4x and 5x to the list.

jayeb replied:
Apparently so: https://gist.github.com/cheeaun/8838438

Also, the new iPhone announced this morning is over 450dpi, so I'm guessing they'll end up with a devicePixelRatio north of 3.

frederickfogerty replied:
👍

frederickfogerty commented on addon/components/imgix-image.js:
```diff
@@ -206,12 +194,14 @@ export default Component.extend(ResizeAware, {
   },
 
-  _debugParams: computed('_width', '_height', '_dpr', function() {
+  _debugParams: computed('width', 'height', function() {
```
The original behaviour was to show `auto` for the width or height values when they weren't set. This means that images without fixed dimensions are shown as `auto x auto`. Does this look sufficient?

jayeb replied:
👍

frederickfogerty commented on README.md:
```diff
@@ -184,6 +206,17 @@ import ImgixCoreJs from 'imgix-core-js';
+
+The largest change in this major version bump is the move to width-based `srcSet` and `sizes` for responsiveness. This has a host of benefits, including better server rendering, better responsiveness, less potential for bugs, and perfomance improvements.
+
+- A `sizes` prop should be added to all usages of ember-cli-imgix. If `sizes` is new to you (or even if it's not), Eric's [seminal article on `srcset` and `sizes`](https://ericportis.com/posts/2014/srcset-sizes/) is highly recommended.
```
👍

frederickfogerty commented on addon/components/imgix-image.js:
```diff
@@ -143,31 +93,37 @@ export default Component.extend(ResizeAware, {
+        theseOptions.h = height;
+      }
+
+      const fixedDimensions = width != null || height != null;
```
This is intended. Let me try explain what would happen if this used `&&`:

For example, say that `height` is set to 200px, and width is unset (auto). Using variable srcSet widths, the src sets generated would be `100x200, 108x200, 117x200, ..., 1500x200`. This means that all the images are loaded at a fixed height, and the aspect ratio of the images is changing. This is probably not what the developer is intending. 

Instead, I imagine the user is intending for the image to be loaded at the source aspect ratio (or at a fixed aspect-ratio, when we implement this). 

The situation is worse when just a width is applied, as it is completely overridden by the variable srcSet generation. Again, probably not what the developer is intending.

Does this make sense? Let me know if I'm missing something here.

jayeb replied:
👍

frederickfogerty commented:
Hey there @Duder-onomy. We feel bad for messing you around like this. We definitely didn't intend to bring your work into this library and then make it obsolete.

We (Imgix) were discussing your thoughts on maintaining a fork offline and had the following question. Would you be open to maintaining the auto-sized version of the component in this library? We were thinking of having it as a separate component (e.g. `imgix-image-fluid`). The benefits of this are:

- you don't have to maintain a fork again (sorry!)
- developers have more options for using this library
- you benefit from any updates we make to the ember-cli-imgix core
- this library benefits from any changes you make in fixing stuff for the second component

It seems like a win-win for both sides. We'll still review PRs and everything in a timely manner, as we've been doing up to now. How does this sound to you?

I'm going to spend a bit of time today trying to put a PR together today for a first-cut version of the component, imitating how React works. We'll see how I go.

Duder-onomy commented:
@frederickfogerty No worries about moving things around. At the end of the day you are correct about srcsets. If I could move TheDyrt over to srcsets instantly I would. 

I am totally down to maintain a component within this lib if imgix is cool with that.  I think it would be preferable as 90% of the code would be the same and I would probably just `extend` the main `imgix-image` component anyhow. 

When this PR lands, we should probably update the demo site, then announce it in the Ember Slack/Discord channels.

frederickfogerty commented:
That's good to hear @Duder-onomy! I'd be keen to hear your thoughts on https://github.com/imgix/ember-cli-imgix/pull/57 as well - it's my first attempt at porting over the same behaviour as what happens in React.
